fix odr violation for std::format with single argument std::chrono::duration

nvartolomei opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [x] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

nvartolomei committed:
utils: remove ostream insertion operator overload for chrono duration

1. This results in ODR violation for fmt since through multiple layers
   of template instantiation it ends up with 2 implementation for
   printing std::chrono::duration. One is this with implicit
    fmt/ostream.h support and the other is provided by fmt/chrono.h.

2. The implementation is faulty. It always prints any durations as
   millisecond units without a unit specifier either.

nvartolomei committed:
config/tests: additional coverage for bounded property

nvartolomei committed:
config: explicit formatting of bounded property examples

We have discovered an ODR violation which causes a ducktape test failure
when running them with a bazel build but not with a cmake build.

Test I tried:
`tests/rptest/tests/cluster_config_test.py::ClusterConfigTest.test_valid_settings`

Errors I get:
`Invalid cpu_profiler_sample_period_ms: '"100ms"' (YAML::TypedBadConversion<std::__2::chrono::duration<long long, std::__2::ratio<1l, 1000l>>> (yaml-cpp: error at line 1, column 1: bad conversion))`

With bazel build the property is rendered as "100ms" while with cmake
build as "100".

Under cmake, the formatting is done by the ostream insertion operator
overload in `src/v/utils/to_string.h`:

```cpp
template<typename Rep, typename Period>
inline std::ostream&
operator<<(std::ostream& o, const std::chrono::duration<Rep, Period>& d) {
    fmt::print(
      o,
      "{}",
      std::chrono::duration_cast<std::chrono::milliseconds>(d).count());
    return o;
}
```

Under bazel build, formatting is handled by formatter specializations of
libfmt "fmt/chrono.h".

The behavior is influenced by the linking order. Godbolt repro
https://godbolt.org/z/EbPEx7n58

In a later commit we'll get rid of the ODR violation but before that we
need to ensure example formatting stays as it was in the binaries we
are currently shipping with cmake to avoid breaking changes and to fix
ducktape tests.

Fortunately, bounded property is used only with a reduced number of
types so we can add handling for them in a relatively simple way.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62120-->
<!--ef5d6daf-6170-4ab4-b5df-628af22c8fd5-->
<!--7a8e276d-d47c-4e84-a4da-d127782cfa98-->
<!--766656c3-ad87-470c-9a78-db91f9af23e7-->
<!--72a94a77-5423-401d-b435-3ec7adc8b953-->
<!--25cc0509-0ade-4b2a-89ae-9ce64392a685-->
<!--title-->
## Retry command for [Build#62120](https://buildkite.com/redpanda/redpanda/builds/62120)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/delete_retention_ms_test.py::DeleteRetentionMsTest.test_create_topic_configurations
tests/rptest/tests/alter_topic_configuration_test.py::AlterTopicConfiguration.test_altering_multiple_topic_configurations
tests/rptest/tests/describe_topics_test.py::DescribeTopicsTest.test_describe_topics_with_documentation_and_types
tests/rptest/tests/topic_creation_test.py::CreateTopicsTest.test_create_topic_with_single_configuration_property
tests/rptest/tests/alter_topic_configuration_test.py::AlterTopicConfiguration.test_altering_topic_configuration@{"property":"delete.retention.ms","value":123456789}
tests/rptest/tests/alter_topic_configuration_test.py::AlterTopicConfiguration.test_set_config_from_describe
tests/rptest/tests/internal_topic_protection_test.py::InternalTopicProtectionTest.kafka_protections_disable_config_test@{"protect_config":"kafka_noproduce_topics"}
tests/rptest/tests/delete_retention_ms_test.py::DeleteRetentionMsTest.test_alter_topic_config
tests/rptest/tests/delete_retention_ms_test.py::DeleteRetentionMsTest.test_set_cluster_config
tests/rptest/tests/mirror_maker_test.py::TestMirrorMakerService.test_simple_end_to_end@{"source_type":"redpanda"}
tests/rptest/tests/alter_topic_configuration_test.py::AlterTopicConfiguration.test_alter_config_does_not_change_replication_factor
tests/rptest/tests/admin_operations_test.py::AdminOperationsTest.test_admin_operations
tests/rptest/tests/delete_retention_ms_test.py::DeleteRetentionMsTest.test_create_topic_configurations_cluster_defaults
tests/rptest/tests/mirror_maker_test.py::TestMirrorMakerService.test_consumer_group_mirroring@{"source_type":"redpanda"}
tests/rptest/tests/alter_topic_configuration_test.py::AlterTopicConfiguration.test_altering_topic_configuration@{"property":"retention.ms","value":360000}
tests/rptest/tests/internal_topic_protection_test.py::InternalTopicProtectionTest.kafka_protections_disable_config_test@{"protect_config":"kafka_nodelete_topics"}
tests/rptest/tests/write_caching_test.py::WriteCachingPropertiesTest.test_properties
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--ef7673be-3d8f-4adc-ab68-574b409dc7b6-->
<!--TEST RESULTS-->
<!--861e17db-6200-4767-af45-85bd2c050c19-->
<!--TEST RESULTS-->
<!--1fe024f6-3fbb-41ce-9943-6a0e31456499-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62120">build#62120</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/62120#01952afa-a320-4a7d-8c8b-dfdf76c68dd9 | FAIL | 0/2 |
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/62120#01952afa-a320-4b0e-b5e6-b306d165167c | FAIL | 0/2 |
| rptest.tests.admin_operations_test.AdminOperationsTest.test_admin_operations | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b55-a2d4-4d18-aa82-c5a8f051c729 | FAIL | 0/20 |
| rptest.tests.alter_topic_configuration_test.AlterTopicConfiguration.test_alter_config_does_not_change_replication_factor | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b55-a2d4-4d18-aa82-c5a8f051c729 | FAIL | 0/20 |
| rptest.tests.alter_topic_configuration_test.AlterTopicConfiguration.test_altering_multiple_topic_configurations | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b55-a2d5-4126-bebc-9c72470baf3b | FAIL | 0/20 |
| rptest.tests.alter_topic_configuration_test.AlterTopicConfiguration.test_altering_topic_configuration.property=delete.retention.ms.value=123456789 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b55-a2d7-4044-ad6a-d42b508a988a | FAIL | 0/20 |
| rptest.tests.alter_topic_configuration_test.AlterTopicConfiguration.test_altering_topic_configuration.property=delete.retention.ms.value=123456789 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b59-b2dd-4071-8d42-36baaf779d09 | FAIL | 0/20 |
| rptest.tests.alter_topic_configuration_test.AlterTopicConfiguration.test_altering_topic_configuration.property=retention.ms.value=360000 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b55-a2d6-4731-b5f7-430aa9089f82 | FAIL | 0/20 |
| rptest.tests.alter_topic_configuration_test.AlterTopicConfiguration.test_set_config_from_describe | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b55-a2d7-4044-ad6a-d42b508a988a | FAIL | 0/20 |
| rptest.tests.alter_topic_configuration_test.AlterTopicConfiguration.test_set_config_from_describe | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b59-b2dd-4071-8d42-36baaf779d09 | FAIL | 0/20 |
| rptest.tests.controller_snapshot_test.ControllerSnapshotTest.test_join_restart_catch_up | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b55-a2d7-4044-ad6a-d42b508a988a | FLAKY | 1/2 |
| rptest.tests.delete_retention_ms_test.DeleteRetentionMsTest.test_alter_topic_config | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b55-a2d4-4d18-aa82-c5a8f051c729 | FAIL | 0/20 |
| rptest.tests.delete_retention_ms_test.DeleteRetentionMsTest.test_alter_topic_config | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b59-b2dd-4071-8d42-36baaf779d09 | FAIL | 0/20 |
| rptest.tests.delete_retention_ms_test.DeleteRetentionMsTest.test_create_topic_configurations | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b55-a2d5-4126-bebc-9c72470baf3b | FAIL | 0/20 |
| rptest.tests.delete_retention_ms_test.DeleteRetentionMsTest.test_create_topic_configurations_cluster_defaults | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b55-a2d6-4731-b5f7-430aa9089f82 | FAIL | 0/20 |
| rptest.tests.delete_retention_ms_test.DeleteRetentionMsTest.test_set_cluster_config | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b55-a2d4-4d18-aa82-c5a8f051c729 | FAIL | 0/20 |
| rptest.tests.delete_retention_ms_test.DeleteRetentionMsTest.test_set_cluster_config | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b59-b2dd-4071-8d42-36baaf779d09 | FAIL | 0/20 |
| rptest.tests.describe_topics_test.DescribeTopicsTest.test_describe_topics_with_documentation_and_types | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b55-a2d5-4126-bebc-9c72470baf3b | FAIL | 0/20 |
| rptest.tests.internal_topic_protection_test.InternalTopicProtectionTest.kafka_protections_disable_config_test.protect_config=kafka_nodelete_topics | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b55-a2d6-4731-b5f7-430aa9089f82 | FAIL | 0/20 |
| rptest.tests.internal_topic_protection_test.InternalTopicProtectionTest.kafka_protections_disable_config_test.protect_config=kafka_noproduce_topics | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b55-a2d7-4044-ad6a-d42b508a988a | FAIL | 0/20 |
| rptest.tests.mirror_maker_test.TestMirrorMakerService.test_consumer_group_mirroring.source_type=redpanda | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b55-a2d6-4731-b5f7-430aa9089f82 | FAIL | 0/20 |
| rptest.tests.mirror_maker_test.TestMirrorMakerService.test_simple_end_to_end.source_type=redpanda | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b55-a2d4-4d18-aa82-c5a8f051c729 | FAIL | 0/20 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=False.with_iceberg=False.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b59-b2dd-4071-8d42-36baaf779d09 | FLAKY | 1/5 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=False.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b59-b2dd-4071-8d42-36baaf779d09 | FLAKY | 1/4 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b59-b2dd-4071-8d42-36baaf779d09 | FLAKY | 1/4 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=False.with_iceberg=True.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b59-b2dd-4071-8d42-36baaf779d09 | FLAKY | 1/4 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b59-b2dd-4071-8d42-36baaf779d09 | FLAKY | 1/4 |
| rptest.tests.topic_creation_test.CreateTopicsTest.test_create_topic_with_single_configuration_property | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b55-a2d5-4126-bebc-9c72470baf3b | FAIL | 0/20 |
| rptest.tests.write_caching_test.WriteCachingPropertiesTest.test_properties | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b55-a2d6-4731-b5f7-430aa9089f82 | FAIL | 0/20 |
| rptest.tests.write_caching_test.WriteCachingPropertiesTest.test_properties | ducktape | https://buildkite.com/redpanda/redpanda/builds/62120#01952b59-b2dd-4071-8d42-36baaf779d09 | FAIL | 0/20 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62131">build#62131</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.mount_unmount_test.MountUnmountIcebergTest.test_simple_remount.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62131#01952d5a-f43f-4b1b-968c-9831773a3da0 | FLAKY | 1/2 |
| rptest.tests.node_pool_migration_test.NodePoolMigrationTest.test_migrating_redpanda_nodes_to_new_pool.balancing_mode=off.test_mode=TestMode.TIRED_STORAGE.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/62131#01952d5e-f6a0-477f-b48a-ff9789092865 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64119">build#64119</a></summary>


| test_id                                                                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_chunk_read_path_test.CloudStorageChunkReadTest.test_read_when_segment_size_smaller_than_chunk_size                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64119#0195f77e-6cc2-459d-b808-f35d2f6096f9 | FLAKY         | 1/2      |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64119#0195f77e-6cc2-418d-8682-bd229ab6889b | FLAKY         | 1/3      |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64119#0195f77e-6cc0-46a8-99ac-954e29be5763 | FLAKY         | 1/5      |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64119#0195f782-488f-41a6-9fa6-a09dccd72b82 | FLAKY         | 1/2      |
| rptest.tests.maintenance_test.MaintenanceCycleTest.test_leader_distribution.use_rpk=False                                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64119#0195f77e-6cc0-46a8-99ac-954e29be5763 | FLAKY         | 1/2      |
</details>
<!--/body-->

nvartolomei committed:
kafka: remove dead code from config_response_utils

nvartolomei committed:
k/server/tests: do not rely on chrono fmt behavior

Redpanda for a long time had it's own formatting for
std::chrono::millisecond via ostream insertion operator overload which
fmtlib used. Now fmt ships with its own implementation for formatting
chrono types (fmt/chrono.h) and end behavior depends on linking order.

In this test we either hard code the strings or call count() on duration
to get a numeric value the formatting of which I don't expect to change.

nvartolomei committed:
kafka: do not rely on fmt for chrono types

We have discovered an ODR violation which causes a ducktape test failure
when running them with a bazel build but not with a cmake build.

Test I tried:
`tests/rptest/tests/cluster_config_test.py::ClusterConfigTest.test_valid_settings`

Errors I get:
`Invalid cpu_profiler_sample_period_ms: '"100ms"' (YAML::TypedBadConversion<std::__2::chrono::duration<long long, std::__2::ratio<1l, 1000l>>> (yaml-cpp: error at line 1, column 1: bad conversion))`

With bazel build the property is rendered as "100ms" while with cmake
build as "100".

Under cmake, the formatting is done by the ostream insertion operator
overload in `src/v/utils/to_string.h`:

```cpp
template<typename Rep, typename Period>
inline std::ostream&
operator<<(std::ostream& o, const std::chrono::duration<Rep, Period>& d) {
    fmt::print(
      o,
      "{}",
      std::chrono::duration_cast<std::chrono::milliseconds>(d).count());
    return o;
}
```

Under bazel build, formatting is handled by formatter specializations of
libfmt "fmt/chrono.h".

The behavior is influenced by the linking order. Godbolt repro
https://godbolt.org/z/EbPEx7n58

In a later commit we will get rid of ostream insertion operator overload
and rely solely on libfmt which formats chrono types with unit suffix.
However, in the kafka responses we have an API contract to serialize
these values without units so do it explicitly.

michael-redpanda reviewed (commented):
Changes look fine.  I would request a rewording of the second to last commit:

> 2. The implementation is garbage. It always prints any durations as
   milliseconds.

dotnwat commented on src/v/config/bounded_property.h:
```diff
@@ -303,7 +304,16 @@ class bounded_property : public property<T> {
+        } else if constexpr (std::is_arithmetic_v<I>) {
+            return fmt::format("{}", guess);
+        } else {
+            static_assert(
+              false, "type does not have a valid example generation strategy");
```
i thought i read somewhere that in c++23 or c++26 this will work like this, but afaik we still need to assert on something that is dependent on T (see base::unsupported_type<T>`).

nvartolomei replied:
This /always/ worked for me, C++17 https://godbolt.org/z/ne1rjnqMs

dotnwat commented on src/v/kafka/server/tests/alter_config_test.cc:
```diff
@@ -1140,8 +1136,7 @@ FIXTURE_TEST(test_incremental_alter_config_remove, alter_config_test_fixture) {
     auto describe_resp = describe_configs(test_tp);
-    assert_property_value(
-      test_tp, "retention.ms", fmt::format("{}", 1234ms), describe_resp);
+    assert_property_value(test_tp, "retention.ms", "1234", describe_resp);
```
oh interesting so this test wasn't robust enough?

nvartolomei replied:
In this unit test, prior to removing ostream overload for std::chrono::duration formatting, we were ending up using it and rendering chrono without units. Because that implementation has been removed but the server implementation continues to rendering unit-less values, as it should otherwise, we have to update the test not to use fmtlib.

The next thing you will probably wonder is if this means that some alter topic config tests were broken as well... After almost 1 hour of digging I have the answer...

No ducktape related to alter topic configs were broken... Because of couple of reasons:
1) log.retention.ms is rendered manually as a number ([ref](https://github.com/redpanda-data/redpanda/blob/6970c2e03f4fb3f6268a810a043a53e2291cbb30/src/v/kafka/server/handlers/configs/config_response_utils.cc#L1103))
2) newer, we do use format for iceberg lag ms (in dev but not in this branch)
  - we don't have a test for it anyway 😞 
  - I added a test locally but it didn't break 😮. The test doesn't break because the fmt templates are instantiated for a `const` argument ([ref](https://github.com/redpanda-data/redpanda/blob/6970c2e03f4fb3f6268a810a043a53e2291cbb30/src/v/kafka/server/handlers/configs/config_response_utils.cc#L65)) and that one doesn't collide with the one which gets generated by src/v/cluster/archival/types.cc with `fmt/chrono.h` include 🤯 
  - you might also wonder what happens if we drop the const qualifier from that method... it still doesn't break the test... Instead, it "fixes" the originally failing test 😮 because now this is the override which "wins" during linking and it is the one from our code rather than the `fmt/chrono.h`

dotnwat commented on src/v/kafka/server/handlers/configs/config_response_utils.cc:
```diff
@@ -11,6 +11,7 @@
 
```
> See previous commits in the series for details.

copy and paste anything here that would be needed to understand this in isolation. for example, this wouldn't work in a `git blame` situation.

dotnwat commented on src/v/utils/to_string.h:
```diff
@@ -43,16 +43,6 @@ std::ostream& operator<<(std::ostream& os, const std::variant<T...>& v) {
     return os;
 }
 
-template<typename Rep, typename Period>
-inline std::ostream&
-operator<<(std::ostream& o, const std::chrono::duration<Rep, Period>& d) {
-    fmt::print(
-      o,
-      "{}",
-      std::chrono::duration_cast<std::chrono::milliseconds>(d).count());
-    return o;
-}
-
```
🔥

dotnwat reviewed (commented):
lgtm

bashtanov commented on src/v/config/bounded_property.h:
```diff
@@ -303,7 +304,16 @@ class bounded_property : public property<T> {
+
+        if constexpr (::detail::
+                        is_specialization_of_v<I, std::chrono::duration>) {
+            return fmt::format("{}", guess.count());
```
This is consistent with `YAML::convert<std::chrono::milliseconds>` from `src/v/config/convert.h` (which is subject to similar ODR violations on yaml-cpp upgrade BTW). And it's not ideal that these two pieces of logic that are meant to be consistent are located so far apart.

What do you think of using something like `YAML::convert<I>::encode(guess)` instead and then dumping it using `Node::Scalar()` without checking individual types? This is to make sure examples provided are consistent with how we parse values if supported types or their parsing rules change.

nvartolomei replied:
I agree with you but as I said in the other thread I didn't want to refactor the too much (config system) in this PR but instead just get rid of one particular problem.

I was thinking the same. We have multiple ways of serialising config values and they aren't necessarily consistent with each other and are far apart which increases the chance of divergence.

Ideally, 1) the config system shouldn't stringify anything, 2) example field should be of the same type as the value, and 3) all serialisation should be done at a pluggable SerDe layer where it's defined only once how a particular value is converted between c++ representation and api via some sort of traits.

bashtanov replied:
If you think we should not entangle these two places by code, maybe cross-reference them with comments?

dotnwat replied:
is there a concern with the code as-is? it seems ok for now we can improve later?

bashtanov replied:
I'd get the comments added now if possible, as further improvements may never happen.

dotnwat replied:
>maybe cross-reference them with comments?

ahh, yes i missed this comment. agree, definitely worth putting comments in for future readers.

bashtanov commented on src/v/kafka/server/handlers/configs/config_response_utils.cc:
```diff
@@ -63,7 +64,11 @@ static void add_config_if_requested(
+    if constexpr (::detail::is_specialization_of_v<T, std::chrono::duration>) {
+        return ssx::sformat("{}", t.count());
+    } else {
+        return ssx::sformat("{}", t);
```
do we assume T is either chorno or numeric?

dotnwat replied:
probably not in full generality, but i think this change is conservative and addresses exactly the case we observed?

nvartolomei replied:
Yes. There are a lot of other cases in the Kafka later and my goal with this PR was to fix only one problem we observe. Different behavior in bazel build vs cmake build.

nvartolomei replied:
In cluster config I explicitly handle chrono and numeric and static assert otherwise.

bashtanov commented:
@nvartolomei thanks for the fix, @dotnwat thanks for tagging me.

BenPope reviewed (dismissed):
LGTM, does it need a backport?

nvartolomei commented:
All feedback addressed. Please take another look.
